uaclient/apt.py

```diff
@@ -560,3 +566,23 @@ def setup_unauthenticated_repo(
         apt_origin,
         "never",
     )
+
+
+def compare_versions(version1: str, version2: str, relation: str) -> bool:
+    """Return True comparing version1 to version2 with the given relation."""
+    try:
+        system.subp(
+            ["dpkg", "--compare-versions", version1, relation, version2]
+        )
+        return True
+    except exceptions.ProcessExecutionError:
+        return False
+
+
+def get_apt_cache_time():
+    cache_time = inf
+    if os.path.exists(APT_UPDATE_SUCCESS_STAMP_PATH):
+        cache_time = os.stat(APT_UPDATE_SUCCESS_STAMP_PATH).st_mtime
+    elif os.path.exists(APT_LISTS_PATH):
+        cache_time = os.stat(APT_LISTS_PATH).st_mtime
```

Comment:
Curious: in what cases would one of these files be present but not the other? (I know nothing about these)

Reply:
Well some versions of APT didn't have the stamp file - and it is just an empty file that APT will touch after it changes the cache. When the file is not there, we can as a fallback see the APT lists - the raw data that APT actually modifies.

I kinda think that in the systems we care about (Ubuntu, Xenial+) the stamp file will always be there... But I preferred to keep the fallback option when I bald-facedly copied this script from the [Ansible codebase](https://github.com/ansible/ansible/blob/12734fa21c08a0ce8c84e533abdc560db2eb1955/lib/ansible/modules/apt.py#L1012-L1022)

Reply:
I see I see. My mind is turning on a way that we could trigger messaging updates based on watching this stamp file rather than on a 12 hour timer... but that's for another PR.